File: js/17298.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const arr = inputs[1].split(' ').map(Number);

const stack = [];
const result = Array(N).fill(-1);

for (let i = 0; i < N; i++) {
  while (stack.length !== 0 && arr[i] > arr[stack[stack.length - 1]]) {
    result[stack[stack.length - 1]] = arr[i];
    stack.pop();
  }
  stack.push(i);
}

console.log(result.join(' '));

File: c++/17298.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

vector<int> v;
stack<int> st;
int result[1000000];

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    int a;
    cin >> a;
    v.push_back(a);
    result[i] = -1;
  }

  for (int i = 0; i < N; i++) {
    while (!st.empty() && v[i] > v[st.top()]) {
      result[st.top()] = v[i];
      st.pop();
    }
    st.push(i);
  }

  for (int i = 0; i < N; i++) {
    cout << result[i] << ' ';
  }

  return 0;
}